File: tests/test_static_pool_equilibrium.py
# ==============================================================================
# Test Name: test_static_pool_equilibrium.py
# Description: End-to-end integration test verifying static pool equilibrium 
#              and spurious current suppression under a constant downward 
#              gravitational field in 3D incompressible Navier-Stokes flow.
# ==============================================================================

# LITERATE TESTING NARRATIVE & MATHEMATICAL GOVERNING EQUATIONS:
# -------------------------------------------------------------
# Experiment Description:
# We configure a closed, rigid 3D rectangular fluid container filled with an 
# incompressible fluid under a constant downward gravitational field (g_y = -9.81 m/s^2). 
# The pressure field is initialized explicitly to satisfy the hydrostatic balance equation:
#
#     p(y) = -\rho \cdot g_y \cdot (y_{\text{top}} - y)
#
# Why We Are Doing It:
# In fractional-step Navier-Stokes solvers, numerical inconsistencies between the pressure 
# gradient discretization and gravity body force terms frequently induce artificial 
# acceleration, causing parasitic currents (spurious velocities) to spontaneously 
# emerge in a fluid that should otherwise remain completely static.
#
# What We Are Trying to Prove:
# We verify that the complete projection pipeline—comprising trial velocity prediction, 
# Poisson pressure solve, and divergence-free velocity correction—maintains exact hydrostatic 
# equilibrium. Specifically, we prove that residual fluid velocities remain strictly bounded 
# below the spatial discretization truncation error threshold:
#
#     \max |v| < 6 \times 10^{-3} \text{ m/s}

import io
import json
import sys
import zipfile
from pathlib import Path

import numpy as np
import pytest

# We import the unmocked main entry point of the Python application wrapper
from src.main import main


def test_static_pool_equilibrium(workspace_folder, monkeypatch):
    """
    Verifies that the Navier-Stokes solver maintains static pool equilibrium 
    and suppresses spurious currents under hydrostatic pressure-gravity balance.
    """
    print("\n================================================================================")
    print("DIAGNOSTIC START: test_static_pool_equilibrium")
    print("================================================================================")

    folder = workspace_folder["folder"]
    input_file = workspace_folder["input_file_name"]
    input_path = Path(folder) / input_file
    output_manifest_name = "static_pool_manifest.json"

    print(f"[1/4] Loading input configuration from: {input_path}")
    with open(input_path, "r", encoding="utf-8") as f:
        input_data = json.load(f)

    # Configure a 9x9x9 Cartesian grid with uniform spacing dx = dy = dz = 1.0 m
    nx, ny, nz = 9, 9, 9
    dx, dy, dz = 1.0, 1.0, 1.0
    input_data["grid"] = {
        "x_min": 0.0,
        "x_max": float(nx * dx),
        "y_min": 0.0,
        "y_max": float(ny * dy),
        "z_min": 0.0,
        "z_max": float(nz * dz),
        "nx": nx,
        "ny": ny,
        "nz": nz,
        "dx": dx,
        "dy": dy,
        "dz": dz
    }

    input_data["simulation_parameters"] = {
        "time_step": 0.001,
        "total_time": 0.001,
        "output_interval": 1
    }

    # Setup closed rigid container mask: boundary walls (mask = -1), active fluid interior (mask = 1)
    total_cells = nx * ny * nz
    mask = [1] * total_cells
    for k in range(nz):
        for j in range(ny):
            for i in range(nx):
                if i == 0 or i == nx - 1 or j == 0 or j == ny - 1 or k == 0 or k == nz - 1:
                    idx = k * (nx * ny) + j * nx + i
                    mask[idx] = -1

    input_data["mask"] = mask

    density = 1000.0
    gravity_y = -9.81
    input_data["fluid_properties"] = {
        "density": density,
        "viscosity": 0.001
    }

    input_data["external_forces"] = {
        "force_vector": [0.0, 0.0, 0.0],
        "gravity_vector": [0.0, gravity_y, 0.0]
    }

    # Initialize explicit hydrostatic pressure distribution: p(y) = -rho * g_y * (y_top - y)
    y_top = float(ny - 1) * dy
    initial_pressure = []
    for k in range(nz):
        for j in range(ny):
            for i in range(nx):
                y_coord = float(j) * dy
                p_val = -density * gravity_y * (y_top - y_coord)
                initial_pressure.append(p_val)

    input_data["initial_conditions"] = {
        "u": [0.0] * total_cells,
        "v": [0.0] * total_cells,
        "w": [0.0] * total_cells,
        "p": initial_pressure
    }

    # Using schema-compliant 'pressure' boundary condition type at the top surface
    input_data["boundary_conditions"] = [
        {"location": "wall", "type": "no-slip", "values": {"u": 0.0, "v": 0.0, "w": 0.0}},
        {"location": "top", "type": "pressure", "values": {"p": 0.0}}
    ]

    with open(input_path, "w", encoding="utf-8") as f:
        json.dump(input_data, f, indent=2)

    cli_args = [
        "main.py",
        "--input_output_folder", folder,
        "--input_file_name", input_file,
        "--output_file_name", output_manifest_name,
    ]
    monkeypatch.setattr(sys, "argv", cli_args)

    print("[2/4] Executing unmocked python wrapper main() for static pool equilibrium evaluation...")
    try:
        main()
    except (RuntimeError, ValueError, OSError, ArithmeticError, SystemExit) as e:
        pytest.fail(f"Static Pool Equilibrium Test Failed: Pipeline threw unexpected exception: {e}")

    print("[3/4] Validating manifest status and extracting ZIP container output...")
    manifest_path = Path(folder) / output_manifest_name
    assert manifest_path.is_file(), f"Manifest file missing at {manifest_path}"

    with open(manifest_path, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    assert manifest["results"]["status"] == "SUCCESS", "Static Pool Test Failed: Pipeline did not complete with SUCCESS status."

    zip_filename = manifest["results"]["zip_filename"]
    zip_path = Path(folder) / zip_filename
    assert zip_path.is_file(), f"ZIP archive missing at {zip_path}"

    print("[4/4] Measuring maximum residual vertical velocity across active fluid domain...")
    with zipfile.ZipFile(zip_path, "r") as zf:
        namelist = zf.namelist()
        step_str = "000001"
        v_name = f"field_v_step_{step_str}.npy"

        assert v_name in namelist, f"Missing snapshot {v_name} in archive."

        v = np.load(io.BytesIO(zf.read(v_name))).reshape((nz, ny, nx))
        mask_arr = np.array(mask).reshape((nz, ny, nx))

        max_v_final = 0.0
        active_fluid_count = 0

        for k in range(nz):
            for j in range(ny):
                for i in range(nx):
                    if mask_arr[k, j, i] == 1:
                        max_v_final = max(max_v_final, abs(v[k, j, i]))
                        active_fluid_count += 1

    assert active_fluid_count > 0, "Error: No active fluid cells found in integration test domain."
    print(f"[debug] max_v_final = {max_v_final}")

    # Assertion: Spurious velocity currents must remain strictly below the truncation error threshold
    assert max_v_final < 6.0e-3, "Static Pool Equilibrium Failure: Spurious currents generated in equilibrium."

    print("DIAGNOSTIC SUCCESS: Static Pool Equilibrium Integration Test validated successfully.")

File: cpp/cpp_integration_tests/test_static_pool_equilibrium.cpp
/**
 * @file test_static_pool_equilibrium.cpp
 * @brief Literate Integration Test Suite for Static Pool Equilibrium 
 *        and Spurious Current Suppression.
 * 
 * LITERATE TESTING NARRATIVE & MATHEMATICAL FORMULATION:
 * ---------------------------------------------------------------------------------
 * Experiment Description:
 * We configure a closed, rigid 3D rectangular fluid container filled with an 
 * incompressible fluid under a constant downward gravitational field (g_y = -9.81 m/s^2). 
 * The pressure field is initialized explicitly to satisfy the hydrostatic balance equation:
 *     p(y) = -rho * g_y * (y_top - y)
 * 
 * Why We Are Doing It:
 * In fractional-step Navier-Stokes solvers, numerical inconsistencies between the pressure 
 * gradient discretization and gravity body force terms frequently induce artificial 
 * acceleration. This flaw causes parasitic currents (spurious velocities) to spontaneously 
 * emerge in a fluid that should otherwise remain completely static.
 * 
 * What We Are Trying to Prove:
 * We aim to verify that the complete projection pipeline—comprising trial velocity prediction, 
 * Poisson pressure solve, and divergence-free velocity correction—maintains exact hydrostatic 
 * equilibrium. Specifically, we prove that residual fluid velocities remain strictly bounded 
 * below the spatial discretization truncation error threshold (max |v| < 6 * 10^-3 m/s).
 * ---------------------------------------------------------------------------------
 */

#include <gtest/gtest.h>
#include <nlohmann/json.hpp>
#include <vector>
#include <cmath>
#include <algorithm>
#include <cassert>
#include "orchestrator.hpp"
#include "predictor.hpp"
#include "pressure_poisson_solver.hpp"
#include "corrector.hpp"
#include "simulation_prestep.hpp"
#include "grid_math.hpp"

using namespace navier_stokes_solver;

TEST(StaticPoolDiagnosticTest, StaticPoolEquilibrium) {
    // We define a 3D cartesian grid of dimensions 9 x 9 x 9 with 
    // uniform mesh spacing dx = dy = dz = 1.0 m.
    int nx = 9;
    int ny = 9;
    int nz = 9;
    double dx = 1.0;
    double dy = 1.0;
    double dz = 1.0;

    GridDimensions dims{nx, ny, nz, dx, dy, dz};
    assert(nx > 0 && ny > 0 && nz > 0);
    assert(dx > 0.0 && dy > 0.0 && dz > 0.0);
    
    // The fluid medium is water, characterized by density rho = 1000.0 kg/m^3 
    // and dynamic viscosity mu = 0.001 Pa*s. The temporal discretization 
    // uses a controlled step size dt = 0.001 s.
    double density = 1000.0; 
    double mu = 0.001;       
    double dt = 0.001;       
    assert(density > 0.0);
    assert(mu >= 0.0);
    assert(dt > 0.0);

    SolverConfig config;
    config.density = density;
    config.max_poisson_iterations = 500;
    config.poisson_tolerance = 1e-12;

    size_t total_cells = static_cast<size_t>(nx) * ny * nz;
    
    // We allocate velocity components (u, v, w) and scalar pressure (p) fields initialized to zero.
    std::vector<double> u(total_cells, 0.0);
    std::vector<double> v(total_cells, 0.0);
    std::vector<double> w(total_cells, 0.0);
    std::vector<double> p(total_cells, 0.0);

    // We configure a closed rigid container. The mask array marks active fluid interior 
    // cells with 1 and boundary wall cells with -1.
    std::vector<int> mask(total_cells, 1);
    for (int k = 0; k < nz; ++k) {
        for (int j = 0; j < ny; ++j) {
            for (int i = 0; i < nx; ++i) {
                size_t idx = get_flat_index(i, j, k, nx, ny);
                if (i == 0 || i == nx - 1 || j == 0 || j == ny - 1 || k == 0 || k == nz - 1) {
                    mask[idx] = -1; // Rigid boundary wall
                }
            }
        }
    }

    // To maintain a static equilibrium under gravity (g_y = -9.81 m/s^2), 
    // the pressure field must balance the body force exactly according to the hydrostatic law:
    //     p(y) = -rho * g_y * (y_top - y)
    double gravity_y = -9.81;
    for (int k = 0; k < nz; ++k) {
        for (int j = 0; j < ny; ++j) {
            for (int i = 0; i < nx; ++i) {
                size_t idx = get_flat_index(i, j, k, nx, ny);
                double y_coord = static_cast<double>(j) * dy;
                double y_top = static_cast<double>(ny - 1) * dy;
                p[idx] = -density * gravity_y * (y_top - y_coord);
            }
        }
    }

    std::vector<double> u_star(total_cells, 0.0);
    std::vector<double> v_star(total_cells, 0.0);
    std::vector<double> w_star(total_cells, 0.0);
    std::vector<double> rhs(total_cells, 0.0);

    std::vector<BoundaryCondition> bc_list;
    BoundaryCondition bc;
    bc.location = "wall";
    bc.type = "velocity";
    bc.u_val = 0.0; bc.v_val = 0.0; bc.w_val = 0.0; bc.scalar_p = 0.0;
    bc_list.push_back(bc);
    assert(!bc_list.empty());

    std::vector<double> gravity = {0.0, gravity_y, 0.0};
    std::vector<double> fx(total_cells, 0.0);
    std::vector<double> fy(total_cells, 0.0);
    std::vector<double> fz(total_cells, 0.0);

    // Execute boundary condition setup prior to evaluating momentum equations.
    execute_pre_step(u, v, w, p, mask, bc_list, nx, ny, nz, false);

    // Compute intermediate trial velocities (u_star) accounting for advection, 
    // diffusion, and external gravity body forces without pressure gradient coupling.
    FluidProperties fluid{mu / density, density};
    compute_trial_velocities(
        dims, fluid, dt,
        u.data(), v.data(), w.data(),
        fx.data(), fy.data(), fz.data(),
        gravity, mask,
        u_star.data(), v_star.data(), w_star.data()
    );

    // The divergence of the trial velocity field acts as the source term for the 
    // pressure correction Poisson equation:
    //     Laplacian(p^(n+1)) = (rho / dt) * div(u_star)
    const double scale = density / dt;
    for (int k = 0; k < nz; ++k) {
        for (int j = 0; j < ny; ++j) {
            for (int i = 0; i < nx; ++i) {
                const size_t idx = get_flat_index(i, j, k, nx, ny);
                if (mask[idx] == 1) {
                    const size_t idx_east  = get_flat_index(i + 1, j, k, nx, ny);
                    const size_t idx_west  = get_flat_index(i - 1, j, k, nx, ny);
                    const size_t idx_north = get_flat_index(i, j + 1, k, nx, ny);
                    const size_t idx_south = get_flat_index(i, j - 1, k, nx, ny);
                    const size_t idx_up    = get_flat_index(i, j, k + 1, nx, ny);
                    const size_t idx_down  = get_flat_index(i, j, k - 1, nx, ny);

                    double dudx = (u_star[idx_east]  - u_star[idx_west])  / (2.0 * dx);
                    double dvdy = (v_star[idx_north] - v_star[idx_south]) / (2.0 * dy);
                    double dwdz = (w_star[idx_up]    - w_star[idx_down])    / (2.0 * dz);

                    rhs[idx] = scale * (dudx + dvdy + dwdz);
                } else {
                    rhs[idx] = 0.0;
                }
            }
        }
    }

    // Solve the Poisson equation iteratively using parallelized Red-Black Gauss-Seidel sweeps.
    solve_poisson_red_black_parallel(
        p, rhs, mask, bc_list,
        nx, ny, nz, dx, dy, dz,
        config.max_poisson_iterations, config.poisson_tolerance,
        config.density, gravity
    );

    // Project the velocity field divergence-free by subtracting the pressure gradient:
    //     u^(n+1) = u_star - (dt / rho) * grad(p^(n+1))
    solve_corrector_parallel(
        u, v, w,
        u_star, v_star, w_star,
        p, mask,
        nx, ny, nz, dx, dy, dz,
        dt, density
    );

    // Measure the maximum residual vertical velocity magnitude across active fluid cells.
    double max_v_final = 0.0;
    for (size_t idx = 0; idx < total_cells; ++idx) {
        if (mask[idx] == 1) {
            max_v_final = std::max(max_v_final, std::abs(v[idx]));
        }
    }

    // Assert that spurious velocity currents remain strictly below the discretization error threshold:
    //     max |v| < 6 * 10^-3 m/s
    assert(max_v_final < 6e-3);
    ASSERT_LT(max_v_final, 6e-3) << "Static Pool Failure: Spurious currents generated in equilibrium.";
}
